### PR TITLE
OkHttp Implementation Change

### PR DESCRIPTION
**Pull request by SizableShrimp**
### Prerequisites
* [x] I have read and understood the [contribution guidelines](CONTRIBUTING.md)
* [x] This pull request follows the code style of the project
* [x] I have tested this feature thoroughly
  * [Proof of testing](https://i.imgur.com/G1cC43P.png) - I have also tested each http method (GET, POST, PUT, PATCH, DELETE) individually

**Issues Fixed:** *hopefully* fixes the issue with bots sometimes getting random DiscordExceptions about Discord not returning a response when experiencing a long period of inactivity

### Changes Proposed in this Pull Request
* Implements the Requests class with OkHttp instead of Apache HttpClient

### CODE REVIEW

**Comment by danthonywalker** [comment]
Remove the Apache HTTPClient dependency(ies)

File: src/main/java/sx/blah/discord/api/internal/Requests.java [hunk 1]
```diff
@@ -56,62 +58,67 @@
...
+	 */
+	public enum RequestMethod {
+		POST, GET, DELETE, PATCH, PUT
 	}
 
 	/**
 	 * A specific HTTP method request type.
 	 */
-	public final class Request {
+	public final class DiscordRequest {
```
Comment (GrandPanda) [inline src/main/java/sx/blah/discord/api/internal/Requests.java]: This class can be private

File: src/main/java/sx/blah/discord/api/internal/Requests.java [hunk 2]
```diff
@@ -56,62 +58,67 @@
...
 
-		POST = new Request(HttpPost.class, client);
-		GET = new Request(HttpGet.class, client);
-		DELETE = new Request(HttpDelete.class, client);
-		PATCH = new Request(HttpPatch.class, client);
-		PUT = new Request(HttpPut.class, client);
+	/**
+	 * The method types available for a request.
+	 */
+	public enum RequestMethod {
```
Comment (GrandPanda) [inline src/main/java/sx/blah/discord/api/internal/Requests.java]: Does this enum need to exist?

File: src/main/java/sx/blah/discord/api/internal/Requests.java [hunk 3]
```diff
@@ -56,62 +58,67 @@
 	 */
 	public static final Requests GENERAL_REQUESTS = new Requests(null);
 
+	/**
+	 * The HTTP client requests are made on.
+	 */
+	private final OkHttpClient HTTP;
```
Comment (GrandPanda) [inline src/main/java/sx/blah/discord/api/internal/Requests.java]: Can this as well as the DiscordRequest fields be renamed to follow convention? (I know you didn't name the other ones)

File: src/main/java/sx/blah/discord/api/internal/Requests.java (hunk 1 above)
Comment (SizableShrimp) [reply src/main/java/sx/blah/discord/api/internal/Requests.java]: I tried changing the access modifier to private and got errors when compiling the jar. I am open to using other access modifiers if applicable.

File: src/main/java/sx/blah/discord/api/internal/Requests.java [hunk 4]
```diff
@@ -56,62 +58,60 @@
...
-		POST = new Request(HttpPost.class, client);
-		GET = new Request(HttpGet.class, client);
-		DELETE = new Request(HttpDelete.class, client);
-		PATCH = new Request(HttpPatch.class, client);
-		PUT = new Request(HttpPut.class, client);
+		HttpLoggingInterceptor httpLogger = new HttpLoggingInterceptor(message -> Discord4J.LOGGER.debug(LogMarkers.API, message)).setLevel(HttpLoggingInterceptor.Level.HEADERS);
+		httpLogger.redactHeader("Set-Cookie");
+		httpLogger.redactHeader("Cookie");
+		httpLogger.redactHeader("Authorization");
+		http = new OkHttpClient.Builder().addInterceptor(httpLogger).build();
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/api/internal/Requests.java]: We will want to test the scalability of this configuration. We may want to tweak `connectionPool()` and/or `dispatcher()`.

File: src/main/java/sx/blah/discord/api/internal/Requests.java [hunk 5]
```diff
@@ -176,7 +176,15 @@ private Request(Class<? extends HttpUriRequest> clazz, DiscordClientImpl client)
 		 */
 		public <T> T makeRequest(String url, Class<T> clazz, BasicNameValuePair... headers) {
 			try {
-				String response = makeRequest(url, headers);
+				Request.Builder builder = new Request.Builder().url(url).header("User-Agent", USER_AGENT);
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/api/internal/Requests.java]: You may wanna turn this into a convenience method

**Comment by NorbiPeti** [comment]
This seemed to work for a long time, then now I'm also getting the same error with this version as well... Now (after some inactivity) all messages the bot tries to send fail, just as before.

```
[19:48:16 ERROR]: Discord4J Internal Exception
java.net.SocketTimeoutException: timeout
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.java:656) ~[?:?]
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.java:664) ~[?:?]
	at okhttp3.internal.http2.Http2Stream.takeHeaders(Http2Stream.java:153) ~[?:?]
	at okhttp3.internal.http2.Http2Codec.readResponseHeaders(Http2Codec.java:131) ~[?:?]
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[?:?]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[?:?]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[?:?]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[?:?]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[?:?]
	at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.java:225) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[?:?]
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:254) ~[?:?]
	at okhttp3.RealCall.execute(RealCall.java:92) ~[?:?]
	at sx.blah.discord.api.internal.Requests$DiscordRequest.request(Requests.java:271) ~[?:?]
	at sx.blah.discord.api.internal.Requests$DiscordRequest.request(Requests.java:235) ~[?:?]
	at sx.blah.discord.api.internal.Requests$DiscordRequest.makeRequest(Requests.java:205) ~[?:?]
	at sx.blah.discord.api.internal.Requests$DiscordRequest.makeRequest(Requests.java:161) ~[?:?]
	at sx.blah.discord.handle.impl.obj.Channel.sendMessage(Channel.java:468) ~[?:?]
	at buttondevteam.discordplugin.DiscordPlugin.lambda$sendMessageToChannel$8(DiscordPlugin.java:423) ~[?:?]
	at sx.blah.discord.util.RequestBuffer$RequestFuture$RequestCallable.call(RequestBuffer.java:286) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at sx.blah.discord.util.RequestBuffer$RequestFuture.run(RequestBuffer.java:259) ~[?:?]
	at sx.blah.discord.util.RequestBuffer$RequestFuture.access$200(RequestBuffer.java:162) ~[?:?]
	at sx.blah.discord.util.RequestBuffer.lambda$request$0(RequestBuffer.java:50) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:844) [?:?]
[19:48:16 WARN]: RequestBuffer handled an uncaught exception!
sx.blah.discord.util.DiscordException: Message was unable to be sent (Discord didn't return a response).
	at sx.blah.discord.handle.impl.obj.Channel.sendMessage(Channel.java:477) ~[?:?]
	at buttondevteam.discordplugin.DiscordPlugin.lambda$sendMessageToChannel$8(DiscordPlugin.java:423) ~[?:?]
	at sx.blah.discord.util.RequestBuffer$RequestFuture$RequestCallable.call(RequestBuffer.java:286) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at sx.blah.discord.util.RequestBuffer$RequestFuture.run(RequestBuffer.java:259) ~[?:?]
	at sx.blah.discord.util.RequestBuffer$RequestFuture.access$200(RequestBuffer.java:162) ~[?:?]
	at sx.blah.discord.util.RequestBuffer.lambda$request$0(RequestBuffer.java:50) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:844) [?:?]
```

**Comment by Shadorc** [comment]
Discord API is having issues right now, maybe the bug is coming from here
https://status.discordapp.com/

**Comment by SizableShrimp** [comment]
Thanks for pointing this out @Shadorc! At the time of writing this pull request, I forgot to consider that the timeout exception is quite literally a timeout and would most likely be related to 
A) discord being down
B) internet of bot being down

_I would go back and add a try/catch if v2 was still a viable option, but I would recommend v3 as the reconnection to gateway and retrying requests to discord is handled a lot better compared to v2._

**Comment by MrMaleficus** [comment]
Any news on this ?

**Comment by danthonywalker** [comment]
Has this proven to resolve the core issue (Discord not returning a response?)

**Comment by NorbiPeti** [comment]
I've been using this version and I haven't seen that issue since.

**Comment by NorbiPeti** [comment]
Not sure why, but after this last commit I got a NoSuchFieldError on HttpLoggingIterceptor.redactHeader(String), but only in production, not on my machine... Anyways I see this got closed, is the 3.x version using the same HTTP library as 2.x or something else? Would the issue this PR was fixing be present if I migrate over?

**Comment by danthonywalker** [comment]
v3 uses Reactor Netty and as far as we can tell there are no issues.

**Pull request closed without merging**